pkg/frontend/querymiddleware/error_caching.go

```diff
@@ -0,0 +1,182 @@
+		return e.next.Do(ctx, request)
+	}
+
+	e.cacheAttempted.Inc()
+	key := e.cacheKey(tenant.JoinTenantIDs(tenantIDs), request)
+	hashedKey := cacheHashKey(key)
+
+	if cachedErr := e.loadErrorFromCache(ctx, key, hashedKey); cachedErr != nil {
+		e.cacheHits.Inc()
+		spanLog.DebugLog(
+			"msg", "returned cached API error",
+			"error_type", cachedErr.Type,
+			"key", key,
+			"hashed_key", hashedKey,
+		)
+
+		return nil, cachedErr
+	}
+
+	res, err := e.next.Do(ctx, request)
+	if err != nil {
+		var apiErr *apierror.APIError
+		if !errors.As(err, &apiErr) {
+			e.cacheStoreSkipped.WithLabelValues(reasonNotAPIError).Inc()
+			return res, err
+		}
+
+		if cacheable, reason := e.isCacheable(apiErr, request, tenantIDs); !cacheable {
+			e.cacheStoreSkipped.WithLabelValues(reason).Inc()
+			spanLog.DebugLog(
+				"msg", "error result from request is not cacheable",
+				"error_type", apiErr.Type,
+				"reason", reason,
+			)
+			return res, err
+		}
+
+		e.storeErrorToCache(key, hashedKey, apiErr)
+	}
+
+	return res, err
+}
+
+func (e *errorCachingHandler) loadErrorFromCache(ctx context.Context, key, hashedKey string) *apierror.APIError {
+	res := e.cache.GetMulti(ctx, []string{hashedKey})
+	if cached, ok := res[hashedKey]; ok {
+		var cachedError CachedError
+		if err := proto.Unmarshal(cached, &cachedError); err != nil {
+			level.Warn(e.logger).Log("msg", "unable to unmarshall cached error", "err", err)
+			return nil
+		}
+
+		if cachedError.GetKey() != key {
+			level.Debug(e.logger).Log(
+				"msg", "cached error key does not match",
+				"expected_key", key,
+				"actual_key", cachedError.GetKey(),
+				"hashed_key", hashedKey,
+			)
+			return nil
+		}
+
+		return apierror.New(apierror.Type(cachedError.Type), cachedError.Message)
+	}
+
+	return nil
+}
+
+func (e *errorCachingHandler) storeErrorToCache(key, hashedKey string, apiErr *apierror.APIError) {
+	bytes, err := proto.Marshal(&CachedError{
+		Key:     key,
+		Type:    string(apiErr.Type),
+		Message: apiErr.Message,
+	})
+
+	if err != nil {
+		level.Warn(e.logger).Log("msg", "unable to marshal cached error", "err", err)
+		return
+	}
+
+	e.cache.SetAsync(hashedKey, bytes, e.ttl)
+}
+
+func (e *errorCachingHandler) cacheKey(tenantID string, r MetricsQueryRequest) string {
+	return fmt.Sprintf("EC:%s:%s:%d:%d:%d", tenantID, r.GetQuery(), r.GetStart(), r.GetEnd(), r.GetStep())
+}
+
+func (e *errorCachingHandler) isCacheable(apiErr *apierror.APIError, req MetricsQueryRequest, tenantIDs []string) (bool, string) {
+	if apiErr.Type != apierror.TypeBadData && apiErr.Type != apierror.TypeExec {
+		return false, reasonNotCacheableError
+	}
+
+	maxCacheFreshness := validation.MaxDurationPerTenant(tenantIDs, e.limits.MaxCacheFreshness)
+	maxCacheTime := int64(model.Now().Add(-maxCacheFreshness))
+	cacheUnalignedRequests := validation.AllTrueBooleansPerTenant(tenantIDs, e.limits.ResultsCacheForUnalignedQueryEnabled)
+
+	return isRequestCachable(req, maxCacheTime, cacheUnalignedRequests, e.logger)
```

Comment:
why do we need to respect step alignment and max cache times? The way I was thinking about this is that we want to prevent the exact same query from running again because it will, almost certainly, fail again.

Reply:
I don't have a particularly good reason for this. It felt correct to use the same criteria for caching errors as we use for caching requests but I can see why that wouldn't be desirable (we keep getting hit by expensive failing queries and don't have any recourse). Do you think we should ignore step alignment here? And do you think we should respect the max cache freshness setting here?

Reply:
I think it's ok to ignore them.

For query results, they make sense because the cached value may become outdated as new samples (possibly out of order) come in. From the errors you listed I don't see any of them becoming less likely as new data comes in.